File: app/llm/openai_provider.py
"""OpenAI provider 구현체.

상용 API 입력 데이터 학습 미사용(no-training) 조건 전제. 계약·약관 변경 시 재확인 필수.

벤더 SDK import은 provider 구현체 파일에서만 허용된다 (CLAUDE.md 아키텍처 규칙 1).
"""

from openai import APIStatusError, AsyncOpenAI, OpenAIError

from app.exceptions import LLMProviderError
from app.llm.provider import (
    DEFAULT_MAX_RETRIES,
    DEFAULT_MAX_TOKENS,
    DEFAULT_TEMPERATURE,
    DEFAULT_TIMEOUT_SECONDS,
    LLMProvider,
    LLMResponse,
)


class OpenAIProvider(LLMProvider):
    """OpenAI Chat Completions API 구현체."""

    name = "openai"

    def __init__(
        self,
        api_key: str,
        model: str = "gpt-4o",
        timeout_seconds: float = DEFAULT_TIMEOUT_SECONDS,
    ) -> None:
        self.model = model
        # 타임아웃·재시도는 SDK 기본값에 맡기지 않고 명시한다 (provider.py 상수 참조).
        self._client = AsyncOpenAI(
            api_key=api_key, timeout=timeout_seconds, max_retries=DEFAULT_MAX_RETRIES
        )

    async def complete(
        self,
        *,
        system: str,
        user: str,
        max_tokens: int = DEFAULT_MAX_TOKENS,
        temperature: float = DEFAULT_TEMPERATURE,
    ) -> LLMResponse:
        """단일 완성 요청. 실패 시 LLMProviderError를 던진다.

        출력 상한은 max_tokens가 아니라 max_completion_tokens로 보낸다 — 설치된 SDK에서
        max_tokens는 deprecated이고 신형 모델에서는 지원되지 않는다.
        """
        try:
            completion = await self._client.chat.completions.create(
                model=self.model,
                max_completion_tokens=max_tokens,
                temperature=temperature,
                messages=[
                    {"role": "system", "content": system},
                    {"role": "user", "content": user},
                ],
            )
        # 예외 메시지에는 벤더명·상태코드/오류 유형까지만 남긴다 — 문서 본문 유출 금지.
        # 원인 체인(from exc)은 유지한다. APM 도입 시 전송 범위를 다시 점검할 것.
        except APIStatusError as exc:
            raise LLMProviderError(f"openai 호출 실패 (HTTP {exc.status_code})") from exc
        except OpenAIError as exc:
            raise LLMProviderError(f"openai 호출 실패 ({type(exc).__name__})") from exc

        if not completion.choices:
            raise LLMProviderError("openai 호출 실패 (빈 choices)")
        choice = completion.choices[0]
        text = choice.message.content
        if not text:
            # 호출 측이 빈 문자열을 정상 변환 결과로 오인하지 않도록 실패로 처리한다.
            raise LLMProviderError("openai 호출 실패 (빈 응답)")

        usage = completion.usage
        return LLMResponse(
            text=text,
            model=completion.model,
            input_tokens=usage.prompt_tokens if usage is not None else 0,
            output_tokens=usage.completion_tokens if usage is not None else 0,
            truncated=choice.finish_reason == "length",
        )

    async def aclose(self) -> None:
        """SDK HTTP 클라이언트를 닫는다."""
        await self._client.close()

File: backend-kotlin/infrastructure/src/main/kotlin/kr/easydoc/infrastructure/llm/OpenAiProvider.kt
package kr.easydoc.infrastructure.llm

import kr.easydoc.core.exceptions.ConfigurationException
import kr.easydoc.core.exceptions.LlmProviderException
import kr.easydoc.core.llm.LlmCompletion
import kr.easydoc.core.llm.LlmFinishReason
import kr.easydoc.core.llm.LlmOptions
import kr.easydoc.core.llm.LlmPrompt
import kr.easydoc.core.llm.LlmProvider
import kr.easydoc.core.security.Secret
import org.springframework.http.MediaType
import org.springframework.http.client.JdkClientHttpRequestFactory
import org.springframework.web.client.RestClient
import org.springframework.web.client.RestClientException
import tools.jackson.core.JacksonException
import tools.jackson.databind.JsonNode
import tools.jackson.databind.json.JsonMapper
import java.net.http.HttpClient
import java.nio.charset.StandardCharsets
import java.time.Duration

const val OPENAI_PROVIDER_NAME: String = "openai"
const val DEFAULT_OPENAI_MODEL: String = "gpt-4.1"

/**
 * [DEFAULT_OPENAI_MODEL](`gpt-4.1`) 이 지원하는 최대 출력 토큰.
 *
 * 출처: OpenAI 공식 모델 문서(2026-09-02 확인) — https://platform.openai.com/docs/models/gpt-4.1
 * ("32,768 max output tokens").
 *
 * **기본 모델만 다룬다.** `easydoc.llm.model` 로 다른 모델을 지정하면 그 모델의 한도는
 * 신뢰성 있게 알 수 없다(`model` 은 자유 문자열이고, 모델별 한도 표는 조용히 낡는다) —
 * [LlmProviderConfiguration] 이 이 상수를 모델 미지정일 때만 검사에 쓰는 이유다.
 */
const val OPENAI_DEFAULT_MODEL_MAX_OUTPUT_TOKENS: Int = 32_768

internal const val OPENAI_BASE_URL: String = "https://api.openai.com"
internal const val OPENAI_RESPONSES_PATH: String = "/v1/responses"

val OPENAI_CONNECT_TIMEOUT: Duration = Duration.ofSeconds(10)
val OPENAI_READ_TIMEOUT: Duration = Duration.ofSeconds(120)

data class OpenAiSettings(
    val apiKey: Secret = Secret.EMPTY,
    val model: String = DEFAULT_OPENAI_MODEL,
    val baseUrl: String = OPENAI_BASE_URL,
    val connectTimeout: Duration = OPENAI_CONNECT_TIMEOUT,
    val readTimeout: Duration = OPENAI_READ_TIMEOUT,
)

/** OpenAI Responses API 어댑터. 요청 저장은 항상 비활성화한다. */
class OpenAiProvider(private val settings: OpenAiSettings) : LlmProvider {
    override val name: String = OPENAI_PROVIDER_NAME

    private val json = JsonMapper.builder().build()

    private val client: RestClient =
        RestClient
            .builder()
            .baseUrl(settings.baseUrl)
            .requestFactory(
                JdkClientHttpRequestFactory(
                    HttpClient.newBuilder().connectTimeout(settings.connectTimeout).build(),
                ).apply { setReadTimeout(settings.readTimeout) },
            ).build()

    override fun toString(): String = "OpenAiProvider(model=${settings.model})"

    override fun complete(
        prompt: LlmPrompt,
        options: LlmOptions,
    ): LlmCompletion {
        if (settings.apiKey.isBlank()) {
            throw ConfigurationException("openai API 키가 설정되지 않았습니다")
        }

        val payload = json.writeValueAsString(requestBody(prompt, options))
        return parse(post(payload))
    }

    private fun requestBody(
        prompt: LlmPrompt,
        options: LlmOptions,
    ) = json.createObjectNode().apply {
        put("model", settings.model)
        put("instructions", prompt.system)
        put("input", prompt.user)
        put("max_output_tokens", options.maxTokens)
        put("store", false)
    }

    private fun post(payload: String): String =
        execute(payload)?.toString(StandardCharsets.UTF_8) ?: throw failure("응답 본문 없음")

    private fun execute(payload: String): ByteArray? =
        try {
            client
                .post()
                .uri(OPENAI_RESPONSES_PATH)
                .header("Authorization", "Bearer ${settings.apiKey.reveal()}")
                .contentType(MediaType.APPLICATION_JSON)
                .body(payload.toByteArray(StandardCharsets.UTF_8))
                .retrieve()
                .onStatus({ status -> status.isError }) { _, response ->
                    throw failure("HTTP ${response.statusCode.value()}")
                }.body(ByteArray::class.java)
        } catch (exc: RestClientException) {
            throw failure(exc::class.java.simpleName)
        }

    private fun parse(raw: String): LlmCompletion {
        val node = readTree(raw)
        val model = requiredText(node, "model")
        val inputTokens = requiredTokenCount(node, "input_tokens")
        val outputTokens = requiredTokenCount(node, "output_tokens")
        val text = extractText(node)

        return LlmCompletion(
            text = text,
            provider = name,
            model = model,
            inputTokens = inputTokens,
            outputTokens = outputTokens,
            finishReason = finishReason(node),
        )
    }

    private fun readTree(raw: String): JsonNode =
        try {
            json.readTree(raw)
        } catch (exc: JacksonException) {
            throw failure("응답 형식 오류: ${exc::class.java.simpleName}")
        }
}

private fun extractText(node: JsonNode): String =
    node
        .path("output")
        .filter { item -> item.path("type").stringValue("") == "message" }
        .flatMap { item -> item.path("content").toList() }
        .filter { content -> content.path("type").stringValue("") == "output_text" }
        .joinToString("") { content -> content.path("text").stringValue("") }

private fun finishReason(node: JsonNode): LlmFinishReason {
    if (hasRefusal(node)) return LlmFinishReason.REFUSAL

    return when (requiredText(node, "status")) {
        "completed" -> {
            LlmFinishReason.END_TURN
        }

        "incomplete" -> {
            when (node.path("incomplete_details").path("reason").stringValue("")) {
                "max_output_tokens" -> LlmFinishReason.MAX_TOKENS
                "content_filter" -> LlmFinishReason.REFUSAL
                else -> LlmFinishReason.OTHER
            }
        }

        "failed", "cancelled" -> {
            throw failure("응답이 완료되지 않았습니다")
        }

        else -> {
            LlmFinishReason.OTHER
        }
    }
}

private fun hasRefusal(node: JsonNode): Boolean =
    node
        .path("output")
        .flatMap { item -> item.path("content").toList() }
        .any { content -> content.path("type").stringValue("") == "refusal" }

private fun requiredText(
    node: JsonNode,
    field: String,
): String = node.path(field).stringValue("").ifEmpty { throw failure("응답에 $field 값이 없습니다") }

private fun requiredTokenCount(
    node: JsonNode,
    field: String,
): Int {
    val value = node.path("usage").path(field)
    if (!value.isIntegralNumber) throw failure("응답 usage.$field 가 정수가 아닙니다")
    val count = value.asInt()
    if (count < 0) throw failure("응답 usage.$field 가 음수입니다")
    return count
}

private fun failure(reason: String) = LlmProviderException("openai 호출 실패 ($reason)")
